### PR TITLE
Move plugin to composer management

### PR DESCRIPTION
**Pull request by LeoColomb**
Using composer for the plugin.

This aims to notably:
* Be able to install the plugin in a composer managed WordPress stack (eg. https://github.com/roots/bedrock)
* Avoid committing Raven files in the repo and be much more flexible with Raven versions
* Autoload smoothly classes when required

🌈

### CODE REVIEW

**Comment by stayallive** [comment]
Hey @LeoColomb,

Can you tell me a bit more. Do I need to run composer install before submitting the plugin to the WordPress plugin directory and zip up the complete vendor folder? Does this even work with the path changes? Please help me understand a bit better, not using any composer based WP install yet. I now need to copy the source everytime, but instead I am going to run composer update everytime I need to update the Sentry SDK, is that how it's suppost to work?

I am not sure I am seeing the benefits of this yet TBH.

**Comment by LeoColomb** [comment]
With pleasure @stayallive! 😃 

> Do I need to run composer install before submitting the plugin to the WordPress plugin directory and zip up the complete vendor folder?

Yes, once. This workflow can easily be included in your `release.sh` with one line, to be automated.
```
composer install --no-dev -o
```

> Does this even work with the path changes?

What do you mean? If composer generate relative or absolute paths?
Everything is relative, definitely no issue with paths.

> not using any composer based WP install yet

OK, let's review basic improvements brought by Composer, for you and your contributors [[1](https://roots.io/using-composer-with-wordpress/)]:
* dependencies are explicitly declared in a single place
* installing and updating is handled by the tool
* project is locked onto specific versions (as before)
* you don’t need to include the actual 3rd party code in your version control repository

Now let's review WordPress with Composer [[2](http://composer.rarst.net/)] [[3](https://wpackagist.org/)] for WordPress users and administrators:
* Much more reliable builds for WordPress when using different environments (eg. dev => staging => production)
* More flexibility regarding plugin management than WordPress GUI
* Avoid committing plugins and themes or having to use git submodules into source control
* Manage WordPress and non-WordPress project libraries with the same tools
* Could eventually be used to manage dependencies between plugins

> I now need to copy the source everytime, but instead I am going to run composer update everytime I need to update the Sentry SDK, is that how it's suppost to work?

Yes, basically, that's the thing! 😉 

> I am not sure I am seeing the benefits of this yet TBH.

Hope it's clear enough now!

**Comment by LeoColomb** [comment]
Oh, I forgot another benefit (but for three files it's not so important), you gain the ability to load lib classes with the [autoloading](https://secure.php.net/manual/fr/language.oop5.autoload.php) benefits.

**Comment by stayallive** [comment]
@LeoColomb thanks for the explanation, there was an issue with the autoloading. And without causing a BC break a classmap will do (for 3 files).

File: composer.json [hunk 1]
```diff
@@ -18,22 +18,12 @@
     "issues": "https://github.com/stayallive/wp-sentry/issues"
   },
   "autoload": {
-    "psr-0": {
-      "WP_Sentry_": "."
-    }
+    "classmap": [
+      "src/"
```
Comment (LeoColomb) [inline composer.json]: 👍

File: composer.json [hunk 2]
```diff
@@ -18,22 +18,12 @@
...
+    ]
   },
   "require": {
     "php": ">=5.4.0",
-    "composer/installers": "~1.0",
```
Comment (LeoColomb) [inline composer.json]: Please keep `composer/installers` as a dependency, it will be a useless composer plugin without it.

File: composer.json [hunk 3]
```diff
@@ -18,22 +18,12 @@
...
     "sentry/sentry": "^1.7.1"
-  },
-  "require-dev": {
-    "dealerdirect/phpcodesniffer-composer-installer": "^0.4.3",
-    "wp-coding-standards/wpcs": "^0.13.1"
-  },
-  "scripts": {
-    "test": [
```
Comment (LeoColomb) [inline composer.json]: Why do you remove test scripts? 🤔

File: src/class-wp-sentry-js-tracker.php [hunk 4]
```diff
@@ -99,10 +99,9 @@ public function get_default_options() {
 	public function on_enqueue_scripts() {
 		wp_enqueue_script(
 			'wp-sentry-raven',
-			plugin_dir_url( WP_SENTRY_PLUGIN_FILE ) . 'raven/js/raven-3.17.0.min.js',
+			plugin_dir_url( WP_SENTRY_PLUGIN_FILE ) . 'public/raven-3.17.0.min.js',
```
Comment (LeoColomb) [inline src/class-wp-sentry-js-tracker.php]: 👍

File: composer.json (hunk 2 above)
Comment (LeoColomb) [reply composer.json]: If you need some examples:
* https://github.com/algolia/algoliasearch-wordpress/blob/0d3c256d3a93acdc134c90a78ccb6e01246e2824/composer.json
* https://github.com/Automattic/jetpack/blob/9c1c52ed99f316d489ae1db539d8d778a325a605/composer.json
* https://github.com/roots/soil/blob/d0e409e70dafff4ceeaa8ef941ee319e350317dd/composer.json

Comment (stayallive) [reply composer.json]: Sorry about that. Will put this one back. Somehow thought it was for phpcs.

File: composer.json (hunk 3 above)
Comment (stayallive) [reply composer.json]: Because phpcs is not used (by my / automated) and the current setup had 200+ errors 👎

File: composer.json (hunk 2 above)
Comment (LeoColomb) [reply composer.json]: OK, thanks! 🎉 
BTW don't forget to publish it on [packagist](https://packagist.org/?) after. 😉
Thanks again!

Comment (stayallive) [reply composer.json]: Yes 👍  Done as 2.1.1/2.1.2. Should be OK now. Thanks for PR'ing this 💯

**Comment by hjanuschka** [comment]
best change ❤️

**Pull request merged**